File: backend/app/schemas/system_optimization.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


# Enums
class MetricType(str, Enum):
    CPU = "CPU"
    MEMORY = "MEMORY"
    DISK = "DISK"
    NETWORK = "NETWORK"
    RESPONSE_TIME = "RESPONSE_TIME"
    THROUGHPUT = "THROUGHPUT"


class AlertCondition(str, Enum):
    GT = "GT"
    LT = "LT"
    EQ = "EQ"
    GTE = "GTE"
    LTE = "LTE"


class AlertSeverity(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class RecommendationType(str, Enum):
    PERFORMANCE = "PERFORMANCE"
    COST = "COST"
    RELIABILITY = "RELIABILITY"
    SCALABILITY = "SCALABILITY"


class EffortLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class RecommendationStatus(str, Enum):
    PENDING = "PENDING"
    APPLIED = "APPLIED"
    DISMISSED = "DISMISSED"
    EXPIRED = "EXPIRED"


class AggregationType(str, Enum):
    AVG = "avg"
    SUM = "sum"
    MAX = "max"
    MIN = "min"


# System Metrics Schemas
class SystemMetricsBase(BaseModel):
    metric_type: MetricType
    metric_name: str = Field(..., min_length=1, max_length=100)
    metric_value: float
    unit: Optional[str] = None
    tags: Optional[Dict[str, Any]] = None
    timestamp: Optional[datetime] = None


class SystemMetricsCreate(SystemMetricsBase):
    pass


class SystemMetricsResponse(SystemMetricsBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Agent Performance Metrics Schemas
class AgentPerformanceMetricsBase(BaseModel):
    agent_id: int
    execution_time_ms: Optional[float] = None
    token_count: Optional[int] = None
    cost: Optional[float] = None
    success_rate: Optional[float] = None
    avg_response_time: Optional[float] = None
    error_count: int = 0
    timestamp: Optional[datetime] = None


class AgentPerformanceMetricsCreate(AgentPerformanceMetricsBase):
    pass


class AgentPerformanceMetricsResponse(AgentPerformanceMetricsBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class AgentPerformanceAggregatedStats(BaseModel):
    """Aggregated statistics for agent performance over a time period"""
    agent_id: int
    total_executions: int
    avg_execution_time_ms: Optional[float] = None
    max_execution_time_ms: Optional[float] = None
    min_execution_time_ms: Optional[float] = None
    total_tokens: Optional[int] = None
    total_cost: Optional[float] = None
    avg_success_rate: Optional[float] = None
    total_errors: int
    avg_response_time_ms: Optional[float] = None


# Optimization Recommendation Schemas
class OptimizationRecommendationBase(BaseModel):
    agent_id: Optional[int] = None
    recommendation_type: RecommendationType
    title: str = Field(..., min_length=1, max_length=200)
    description: str
    current_value: Optional[str] = None
    recommended_value: Optional[str] = None
    impact_score: float = Field(..., ge=0, le=1)
    effort_level: EffortLevel
    status: RecommendationStatus = RecommendationStatus.PENDING


class OptimizationRecommendationCreate(OptimizationRecommendationBase):
    pass


class OptimizationRecommendationResponse(OptimizationRecommendationBase):
    id: int
    created_at: datetime
    expires_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class RecommendationApplyRequest(BaseModel):
    apply_changes: bool = False
    auto_approve: bool = False
    notes: Optional[str] = None


# Performance Baseline Schemas
class PerformanceBaselineBase(BaseModel):
    agent_id: Optional[int] = None
    metric_type: MetricType
    baseline_value: float
    threshold_value: Optional[float] = None
    measurement_window: int = Field(default=3600, ge=60)


class PerformanceBaselineCreate(PerformanceBaselineBase):
    pass


class PerformanceBaselineResponse(PerformanceBaselineBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PerformanceBaselineUpdate(BaseModel):
    baseline_value: Optional[float] = None
    threshold_value: Optional[float] = None
    measurement_window: Optional[int] = None


# Alert Rule Schemas
class AlertRuleBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    metric_type: MetricType
    condition: AlertCondition
    threshold_value: float
    severity: AlertSeverity
    is_active: bool = True
    cooldown_period: int = Field(default=300, ge=0)


class AlertRuleCreate(AlertRuleBase):
    pass


class AlertRuleResponse(AlertRuleBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class AlertRuleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    metric_type: Optional[MetricType] = None
    condition: Optional[AlertCondition] = None
    threshold_value: Optional[float] = None
    severity: Optional[AlertSeverity] = None
    is_active: Optional[bool] = None
    cooldown_period: Optional[int] = None


class AlertHistoryResponse(BaseModel):
    id: int
    alert_rule_id: int
    triggered_value: float
    triggered_at: datetime
    acknowledged: bool
    acknowledged_at: Optional[datetime] = None
    acknowledged_by: Optional[str] = None
    notes: Optional[str] = None

    class Config:
        from_attributes = True


# Metrics Query Schemas
class MetricsQueryRequest(BaseModel):
    metric_types: Optional[List[MetricType]] = None
    metric_names: Optional[List[str]] = None
    start_time: datetime
    end_time: datetime
    aggregation: AggregationType = AggregationType.AVG
    tags: Optional[Dict[str, Any]] = None
    limit: int = Field(default=1000, ge=1, le=10000)


class MetricsQueryResponse(BaseModel):
    """Response for metrics query with time-series data"""
    metric_type: MetricType
    metric_name: str
    aggregation: str
    start_time: datetime
    end_time: datetime
    data_points: List[Dict[str, Any]]  # List of {timestamp, value} objects
    statistics: Optional[Dict[str, float]] = None  # {avg, sum, max, min, count}


# System Health Schemas
class SystemHealthStatus(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class SystemHealthResponse(BaseModel):
    status: SystemHealthStatus
    metrics: Dict[str, Any]
    last_updated: datetime
    active_alerts: int = 0
    pending_recommendations: int = 0


# Dashboard Summary Schemas
class DashboardSummary(BaseModel):
    system_health: SystemHealthResponse
    recent_metrics: List[SystemMetricsResponse]
    top_recommendations: List[OptimizationRecommendationResponse]
    active_alerts: List[AlertRuleResponse]

File: frontend/src/types/systemOptimization.ts
// System Optimization Types
// TypeScript interfaces matching backend schemas for system optimization and performance tuning

// Enums (exported as string literal types for TypeScript)
export type MetricType =
    | 'CPU'
    | 'MEMORY'
    | 'DISK'
    | 'NETWORK'
    | 'RESPONSE_TIME'
    | 'THROUGHPUT';

export type AlertCondition = 'GT' | 'LT' | 'EQ' | 'GTE' | 'LTE';

export type AlertSeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type RecommendationType = 'PERFORMANCE' | 'COST' | 'RELIABILITY' | 'SCALABILITY';

export type EffortLevel = 'LOW' | 'MEDIUM' | 'HIGH';

export type RecommendationStatus = 'PENDING' | 'APPLIED' | 'DISMISSED' | 'EXPIRED';

export type AggregationType = 'avg' | 'sum' | 'max' | 'min';

export type SystemHealthStatus = 'healthy' | 'degraded' | 'unhealthy';

// ==================== System Metrics ====================

export interface SystemMetrics {
    id: number;
    metric_type: MetricType;
    metric_name: string;
    metric_value: number;
    unit?: string;
    tags?: Record<string, unknown>;
    timestamp: string;
    created_at: string;
}

export interface SystemMetricsCreate {
    metric_type: MetricType;
    metric_name: string;
    metric_value: number;
    unit?: string;
    tags?: Record<string, unknown>;
    timestamp?: string;
}

// ==================== Agent Performance Metrics ====================

export interface AgentPerformanceMetrics {
    id: number;
    agent_id: number;
    execution_time_ms?: number;
    token_count?: number;
    cost?: number;
    success_rate?: number;
    avg_response_time?: number;
    error_count: number;
    timestamp: string;
    created_at: string;
}

export interface AgentPerformanceMetricsCreate {
    agent_id: number;
    execution_time_ms?: number;
    token_count?: number;
    cost?: number;
    success_rate?: number;
    avg_response_time?: number;
    error_count?: number;
    timestamp?: string;
}

export interface AgentPerformanceAggregatedStats {
    agent_id: number;
    total_executions: number;
    avg_execution_time_ms?: number;
    max_execution_time_ms?: number;
    min_execution_time_ms?: number;
    total_tokens?: number;
    total_cost?: number;
    avg_success_rate?: number;
    total_errors: number;
    avg_response_time_ms?: number;
}

// ==================== Optimization Recommendations ====================

export interface OptimizationRecommendation {
    id: number;
    agent_id?: number;
    recommendation_type: RecommendationType;
    title: string;
    description: string;
    current_value?: string;
    recommended_value?: string;
    impact_score: number;
    effort_level: EffortLevel;
    status: RecommendationStatus;
    created_at: string;
    expires_at?: string;
}

export interface OptimizationRecommendationCreate {
    agent_id?: number;
    recommendation_type: RecommendationType;
    title: string;
    description: string;
    current_value?: string;
    recommended_value?: string;
    impact_score: number;
    effort_level: EffortLevel;
    expires_at?: string;
}

export interface RecommendationApplyRequest {
    apply_changes?: boolean;
    auto_approve?: boolean;
    notes?: string;
}

// ==================== Performance Baselines ====================

export interface PerformanceBaseline {
    id: number;
    agent_id?: number;
    metric_type: MetricType;
    baseline_value: number;
    threshold_value?: number;
    measurement_window: number;
    created_at: string;
    updated_at: string;
}

export interface PerformanceBaselineCreate {
    agent_id?: number;
    metric_type: MetricType;
    baseline_value: number;
    threshold_value?: number;
    measurement_window?: number;
}

export interface PerformanceBaselineUpdate {
    baseline_value?: number;
    threshold_value?: number;
    measurement_window?: number;
}

// ==================== Alert Rules ====================

export interface AlertRule {
    id: number;
    name: string;
    description?: string;
    metric_type: MetricType;
    condition: AlertCondition;
    threshold_value: number;
    severity: AlertSeverity;
    is_active: boolean;
    cooldown_period: number;
    created_at: string;
}

export interface AlertRuleCreate {
    name: string;
    description?: string;
    metric_type: MetricType;
    condition: AlertCondition;
    threshold_value: number;
    severity: AlertSeverity;
    is_active?: boolean;
    cooldown_period?: number;
}

export interface AlertRuleUpdate {
    name?: string;
    description?: string;
    metric_type?: MetricType;
    condition?: AlertCondition;
    threshold_value?: number;
    severity?: AlertSeverity;
    is_active?: boolean;
    cooldown_period?: number;
}

export interface AlertHistory {
    id: number;
    alert_rule_id: number;
    triggered_value: number;
    triggered_at: string;
    acknowledged: boolean;
    acknowledged_at?: string;
    acknowledged_by?: string;
    notes?: string;
}

export interface ActiveAlert {
    alert_rule_id: number;
    alert_rule_name: string;
    triggered_value: number;
    triggered_at: string;
    severity: AlertSeverity;
    condition: string;
}

// ==================== Metrics Query ====================

export interface MetricsQueryRequest {
    metric_types?: MetricType[];
    metric_names?: string[];
    start_time: string;
    end_time: string;
    aggregation: AggregationType;
    tags?: Record<string, unknown>;
    limit?: number;
}

export interface MetricsQueryResponse {
    metric_type: MetricType;
    metric_name: string;
    aggregation: string;
    start_time: string;
    end_time: string;
    data_points: Array<{ timestamp: string; value: number }>;
    statistics?: {
        avg: number;
        sum: number;
        max: number;
        min: number;
        count: number;
    };
}

// ==================== System Health ====================

export interface SystemHealthResponse {
    status: SystemHealthStatus;
    metrics: Record<string, {
        latest: number;
        avg: number;
        max: number;
    }>;
    last_updated: string;
    active_alerts: number;
    pending_recommendations: number;
}

// ==================== Dashboard Summary ====================

export interface DashboardSummary {
    system_health: SystemHealthResponse;
    recent_metrics: Array<{
        id: number;
        metric_type: MetricType;
        metric_name: string;
        metric_value: number;
        timestamp: string;
        unit?: string;
    }>;
    top_recommendations: Array<{
        id: number;
        title: string;
        recommendation_type: RecommendationType;
        impact_score: number;
        effort_level: EffortLevel;
        status: RecommendationStatus;
        description: string;
        created_at: string;
    }>;
    active_alerts: ActiveAlert[];
    generated_at: string;
}

// ==================== Chart Data Types ====================

export interface TimeSeriesDataPoint {
    timestamp: string;
    value: number;
}

export interface TimeSeriesChartData {
    metric_type: MetricType;
    metric_name: string;
    data: TimeSeriesDataPoint[];
    color?: string;
}

export interface MultiSeriesChartData {
    title: string;
    series: Array<{
        name: string;
        data: TimeSeriesDataPoint[];
        color?: string;
    }>;
}

export interface BarChartData {
    label: string;
    value: number;
    color?: string;
}

export interface HeatMapData {
    x: string;
    y: string;
    value: number;
}

// ==================== Filter Types ====================

export interface MetricsFilter {
    metric_types?: MetricType[];
    start_time?: string;
    end_time?: string;
    aggregation?: AggregationType;
}

export interface RecommendationsFilter {
    agent_id?: number;
    recommendation_type?: RecommendationType;
    status?: RecommendationStatus;
    impact_score_min?: number;
}

export interface AlertsFilter {
    metric_type?: MetricType;
    severity?: AlertSeverity;
    is_active?: boolean;
}

export interface DateRange {
    start: Date;
    end: Date;
}

// ==================== API Response Types ====================

export interface ApiResponse<T> {
    data: T;
    success: boolean;
    message?: string;
    error?: string;
}

export interface PaginatedResponse<T> {
    data: T[];
    total: number;
    page: number;
    limit: number;
    has_more: boolean;
}